### PR TITLE
fix(pageserver): ensure upload happens after delete

### PR DESCRIPTION
**Pull request by skyzh**
## Problem

Follow up of https://github.com/neondatabase/neon/pull/9682, that patch didn't fully address the problem: what if shutdown fails due to whatever reason and then we reattach the tenant? Then we will still remove the future layer. The underlying problem is that the fix for #5878 gets voided because of the generation optimizations.

Of course, we also need to ensure that delete happens after uploads, but note that we only schedule deletes when there are no ongoing upload tasks, so that's fine.

## Summary of changes

* Add a test case to reproduce the behavior (by changing the original test case to attach the same generation).
* If layer upload happens after the deletion, drain the deletion queue before uploading.
  * If blocked_deletion is enabled, directly remove it from the blocked_deletion queue.
* Local fs backend fix to avoid race between deletion and preload.
* test_emergency_mode does not need to wait for uploads (and it's generally not possible to wait for uploads).
* ~~Optimize deletion executor to skip validation if there are no files to delete.~~ this doesn't work

### CODE REVIEW

**Comment by skyzh** [comment]
I have no idea why https://github.com/neondatabase/neon/pull/9844/commits/9e7fa9802452c7ab5f511ae4ecc8bf764abf62cc will cause a lot of test failures but after reverting everything seems so good.

File: pageserver/src/tenant/remote_timeline_client.rs [hunk 1]
```diff
@@ -1921,7 +1934,59 @@ impl RemoteTimelineClient {
             }
 
             let upload_result: anyhow::Result<()> = match &task.op {
-                UploadOp::UploadLayer(ref layer, ref layer_metadata) => {
+                UploadOp::UploadLayer(ref layer, ref layer_metadata, last_op) => {
```
Comment (jcsp) [inline pageserver/src/tenant/remote_timeline_client.rs:1937]: `last_op` is a confusing name here: in the code path that inserts this operation we call this `mode`?

Comment (skyzh) [reply pageserver/src/tenant/remote_timeline_client.rs:1937]: ahh, a commit is lost, I remembered I renamed this

File: pageserver/src/tenant/remote_timeline_client.rs [hunk 2]
```diff
@@ -1921,7 +1934,59 @@ impl RemoteTimelineClient {
...
+                                        } else {
+                                            // keep the layer
+                                            true
+                                        }
+                                    });
+                                }
+                            }
+                            if detected {
+                                info!(
+                                    "upload_queue_reordering: cancelled blocked deletion of layer {} at gen {:?}",
```
Comment (jcsp) [inline pageserver/src/tenant/remote_timeline_client.rs:1961]: `upload_queue_reordering: ` is a bit abstract (we're not reordering anything here, just reasoning about whether it is safe to do so).  I would either drop the prefix, or make it "perform_upload_task:" just to make it obvious where the callsite is.

File: pageserver/src/tenant/remote_timeline_client.rs (hunk 1 above)
Comment (skyzh) [reply pageserver/src/tenant/remote_timeline_client.rs]: did a force push and the missing commit shows up in GitHub https://github.com/neondatabase/neon/pull/9844/commits/78c02f32459e4cfbeaebd0f00568889941119a0d

File: pageserver/src/tenant/remote_timeline_client.rs (hunk 2 above)
Comment (skyzh) [reply pageserver/src/tenant/remote_timeline_client.rs]: https://github.com/neondatabase/neon/pull/9844/commits/2a7d00296edc3c7adea82227080c8bbeadb7688e

**Comment by jcsp** [comment]
I see how this PR addresses the case of deleting+recreating the same image layer within the lifetime of a `Timeline`, but what about the case where some earlier shutdown went wrong and we start a new Timeline while some deletions are still in the global deletion queue?  Don't we still need a change similar to https://github.com/neondatabase/neon/pull/9796 to make it flush the deletion queue on start?

**Comment by skyzh** [comment]
> I see how this PR addresses the case of deleting+recreating the same image layer within the lifetime of a `Timeline`, but what about the case where some earlier shutdown went wrong and we start a new Timeline while some deletions are still in the global deletion queue? Don't we still need a change similar to #9796 to make it flush the deletion queue on start?

Yes; in theory, this could happen, but it has not been observed yet, and I assume it is rarer to happen, and we can fix it in another patch. The problem I would like to address is the race between future layer deletion and the upload of the same layer. Both operations happen _after_ the tenant reattach.

**Comment by skyzh** [comment]
... and if I'm going to fix it, I will populate the `recently_deleted` hashset with the initial deletion list.

**Comment by jcsp** [comment]
>problem I would like to address is the race between future layer deletion and the upload of the same layer. Both operations happen after the tenant reattach.

Makes sense

**Review by jcsp — APPROVED** [review]

**Pull request merged**